### ISSUE TITLE
Bug: Requests are executed twice

### ISSUE THREAD

**Issue by zirkelc** [Shopify/shopify-app-js#1048]
# Issue summary

Since upgrading to `@shopify/shopify-app-remix v3.0.0` and `@shopify/shopify-api v11.0.0`, I noticed that all Shopify requests are executed twice. I first noticed it in my own app, but I could reproduce it in the template app created by Shopify CLI:

<img width="1224" alt="image" src="https://github.com/Shopify/shopify-app-js/assets/950244/8c789feb-9311-4d63-9faa-e72c7baacbce">

Here is also a more detailed debug log:
```sh
15:37:54 │ remix         │ [shopify-app/INFO] Authenticating admin request
15:37:54 │remix        │ [shopify-app/DEBUG] Attempting to authenticate session token | {sessionToken: {"search":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczpcL1wvdmV5bm91LXRlc3QubXlzaG9waWZ5LmNvbVwvYWRtaW4iLCJkZXN0IjoiaHR0cHM6XC9cL3ZleW5vdS10ZXN0Lm15c2hvcGlmeS5jb20iLCJhdWQiOiJmMGRkZWFiZjU0MGQyOTA1NzZhNWQ3ZGI3MjI2NDg4MSIsInN1YiI6Ijg0MjY3MzY4NjAyIiwi
                      ZXhwIjoxNzE4Mjg1OTMyLCJuYmYiOjE3MTgyODU4NzIsImlhdCI6MTcxODI4NTg3MiwianRpIjoiY2VlMTVhODgtM2RlNS00YWQ3LTg2MTgtOWI4ODI2Y2YxZWEzIiwic2lkIjoiOWM1NDYxNDUtZTBiOC00MTk2LWFkOTctNjc3MjJjOWQ0YjliIiwic2lnIjoiMWI5MmJlZjAxODMxZTQwYThkMjdjMmNiYzg3NTM0MzZmZGQwZDViYWYwODQ4MDY2YjQzNDAxNDFhZWZmYWIzNyJ9.OTJg6Wlajkg1XkwT1aa0T2fgvGh-Llu6jAgBlvRsJLY"}}
15:37:54 │ remix         │ [shopify-app/DEBUG] Validating session token
15:37:54 │ remix         │ [shopify-app/INFO] Authenticating admin request
15:37:54 │remix        │ [shopify-app/DEBUG] Attempting to authenticate session token | {sessionToken: {"search":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczpcL1wvdmV5bm91LXRlc3QubXlzaG9waWZ5LmNvbVwvYWRtaW4iLCJkZXN0IjoiaHR0cHM6XC9cL3ZleW5vdS10ZXN0Lm15c2hvcGlmeS5jb20iLCJhdWQiOiJmMGRkZWFiZjU0MGQyOTA1NzZhNWQ3ZGI3MjI2NDg4MSIsInN1YiI6Ijg0MjY3MzY4NjAyIiwi
                      ZXhwIjoxNzE4Mjg1OTMyLCJuYmYiOjE3MTgyODU4NzIsImlhdCI6MTcxODI4NTg3MiwianRpIjoiY2VlMTVhODgtM2RlNS00YWQ3LTg2MTgtOWI4ODI2Y2YxZWEzIiwic2lkIjoiOWM1NDYxNDUtZTBiOC00MTk2LWFkOTctNjc3MjJjOWQ0YjliIiwic2lnIjoiMWI5MmJlZjAxODMxZTQwYThkMjdjMmNiYzg3NTM0MzZmZGQwZDViYWYwODQ4MDY2YjQzNDAxNDFhZWZmYWIzNyJ9.OTJg6Wlajkg1XkwT1aa0T2fgvGh-Llu6jAgBlvRsJLY"}}
15:37:54 │ remix         │ [shopify-app/DEBUG] Validating session token
15:37:54 │remix        │ [shopify-app/DEBUG] Session token is valid | {payload: {"iss":"https://veynou-test.myshopify.com/admin","dest":"https://veynou-test.myshopify.com","aud":"f0ddeabf540d290576a5d7db72264881","sub":"84267368602","exp":1718285932,"nbf":1718285872,"iat":1718285872,"jti":"cee15a88-3de5-4ad7-8618-9b8826cf1ea3","sid":"9c546145-e0b8-4196-ad97-677
                      22c9d4b9b","sig":"1b92bef01831e40a8d27c2cbc8753436fdd0d5baf0848066b4340141aeffab37"}}
15:37:54 │remix        │ [shopify-app/DEBUG] Session token is valid | {payload: {"iss":"https://veynou-test.myshopify.com/admin","dest":"https://veynou-test.myshopify.com","aud":"f0ddeabf540d290576a5d7db72264881","sub":"84267368602","exp":1718285932,"nbf":1718285872,"iat":1718285872,"jti":"cee15a88-3de5-4ad7-8618-9b8826cf1ea3","sid":"9c546145-e0b8-4196-ad97-677
                      22c9d4b9b","sig":"1b92bef01831e40a8d27c2cbc8753436fdd0d5baf0848066b4340141aeffab37"}}
15:37:54 │ remix         │ [shopify-app/DEBUG] Session token is valid | {shop: veynou-test.myshopify.com, payload: [object Object]}
15:37:54 │ remix         │ [shopify-app/DEBUG] Session token is valid | {shop: veynou-test.myshopify.com, payload: [object Object]}
15:37:54 │ remix         │ [shopify-app/DEBUG] Loading session from storage | {sessionId: offline_veynou-test.myshopify.com}
15:37:54 │ remix         │ [shopify-app/DEBUG] Loading session from storage | {sessionId: offline_veynou-test.myshopify.com}
15:37:54 │ remix         │ [shopify-app/INFO] No valid session found
15:37:54 │ remix         │ [shopify-app/INFO] Requesting offline access token
15:37:54 │ remix         │ [shopify-app/INFO] No valid session found
15:37:54 │ remix         │ [shopify-app/INFO] Requesting offline access token
15:37:55 │ remix         │ [shopify-api/INFO] Creating new session | {shop: veynou-test.myshopify.com, isOnline: false}
15:37:55 │ remix         │ [shopify-app/INFO] Running afterAuth hook
15:37:55 │ remix         │ -> afterAuth
15:37:55 │ remix         │ [shopify-api/INFO] Registering webhooks | {shop: veynou-test.myshopify.com}
15:37:55 │ remix         │ [shopify-app/DEBUG] Request is valid, loaded session from session token | {shop: veynou-test.myshopify.com, isOnline: false}
15:37:55 │ remix         │ [shopify-api/INFO] Creating new session | {shop: veynou-test.myshopify.com, isOnline: false}
15:37:55 │ remix         │ [shopify-app/DEBUG] Request is valid, loaded session from session token | {shop: veynou-test.myshopify.com, isOnline: false}
15:37:55 │ remix         │ [shopify-api/DEBUG] Existing topics: [] | {shop: veynou-test.myshopify.com}
15:37:55 │ remix         │ [shopify-api/DEBUG] Running webhook mutation | {topic: APP_UNINSTALLED, operation: create}
15:37:56 │ remix         │ [shopify-app/DEBUG] Registered webhook | {topic: APP_UNINSTALLED, shop: veynou-test.myshopify.com, operation: create}
```

Before opening this issue, I have:

- [X] Upgraded to the latest version of the relevant packages
  - `@shopify/*` package and version:
 ```sh
    "@shopify/app-bridge-react": "^4.1.2",
    "@shopify/polaris": "^12.0.0",
    "@shopify/shopify-api": "^11.0.0",
    "@shopify/shopify-app-remix": "^3.0.0",
    "@shopify/shopify-app-session-storage-prisma": "^5.0.0",
``` 
  - Node version: v20.13.1
  - Operating system: macOS 14.1 (23B74)
- [X] Set `{ logger: { level: LogSeverity.Debug } }` in my configuration, when applicable
- [X] Found a reliable way to reproduce the problem that indicates it's a problem with the package
- [X] Looked for similar issues in this repository
- [X] Checked that this isn't an issue with a Shopify API
  - If it is, please create a post in the [Shopify community forums](https://community.shopify.com/c/partners-and-developers/ct-p/appdev) or report it to [Shopify Partner Support](https://help.shopify.com/en/support/partners/org-select)

## Expected behavior

Each request should be executed once.

## Actual behavior

Each request is executed multiple times.

## Steps to reproduce the problem

1. create new app with Shopify CLI `@shopify/cli/3.61.0 darwin-arm64 node-v20.13.1`: `shopify app init`
2. start app locally: `shopify app dev`
3. install app in test store

**Comment by zirkelc** [comment]
I have to row back on this issue, the logs were a bit misleading from actual problem. I investigated the Shopify implementation and found the root cause. When I upgraded to v3 of `shopify-app-remix` and `shopify-api`, I also switched to the managed installation. The managed installation uses a new auth strategy [`TokenExchangeStrategy`](https://github.com/Shopify/shopify-app-js/blob/7045fe4e3a9c39b858ca8e8337978ee42d496b82/packages/apps/shopify-app-remix/src/server/authenticate/admin/strategies/token-exchange.ts) that calls the `afterAuth` hook from this method:
https://github.com/Shopify/shopify-app-js/blob/7045fe4e3a9c39b858ca8e8337978ee42d496b82/packages/apps/shopify-app-remix/src/server/authenticate/admin/strategies/token-exchange.ts#L159-L172 

The [`IdempotentPromiseHandler`](https://github.com/Shopify/shopify-app-js/blob/7045fe4e3a9c39b858ca8e8337978ee42d496b82/packages/apps/shopify-app-remix/src/server/authenticate/helpers/idempotent-promise-handler.ts) seems to de-duplicate multiple `afterAuth` calls for the same session. 

Now the real problem was how I created the `shopifyApp()` until now. I'm running on a AWS Lambda and serve multiple different custom apps on the same handler. So the Shopify config (api key, secret, scope, app url, etc.) differs from request to request and is not statically available. That means I need to pull some information out of the request and initialize a new `shopifyApp()` object for this request. Furthermore, I didn't store this object on the global scope, instead I implemented it as a getter function `getShopifyApp()` which is called from the various routes (React components) that need it. That means on request which goes through multiple routes may create multiple `shopifyApp()` objects through `getShopifyApp()`.

This worked fine in the past, because the [`AuthCodeFlowStrategy`](https://github.com/Shopify/shopify-app-js/blob/7045fe4e3a9c39b858ca8e8337978ee42d496b82/packages/apps/shopify-app-remix/src/server/authenticate/admin/strategies/auth-code-flow.ts) didn't call the `afterAuth` twice. Probably because the `/auth/callback` route wasn't handled by an actual React component. 
However, the new managed installation through `TokenExchangeStrategy` works with the de-duplication mechanism which is saved as local state on the `shopifyApp()`object. That means multiple routes (React components) create multiple `shopifyApp()` objects each with their own local state.

---

I will now refactor my lambda handler to only create one "shopifyApp() object per request and store it either globally or in the request context. Just wanted to give a thorough update in case anyone has made the same mistake I did 😄.
